File: app/src/main/java/com/diamond/appcliente/viewmodel/GestionarBarberoViewModel.java
package com.diamond.appcliente.viewmodel;

import android.content.Context;

import com.diamond.appcliente.api.ApiClient;
import com.diamond.appcliente.api.AuthApiService;
import com.diamond.appcliente.dto.barbero.BarberoDto;
import com.diamond.appcliente.dto.barbero.BarberoRequest;
import com.diamond.appcliente.dto.barbero.BarberoResponse;
import com.diamond.appcliente.dto.barbero.BarberoSimpleResponse;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class GestionarBarberoViewModel {

    public interface BarberoCallback {
        void onSuccess(List<BarberoDto> barberos);
        void onError(String mensaje);
    }

    public interface BarberoOperacionCallback {
        void onSuccess(String mensaje);
        void onError(String mensaje);
    }

    public interface ActualizarCallback {
        void onSuccess(String mensaje);
        void onError(String mensaje);
    }

    public void obtenerBarberos(Context context, BarberoCallback callback) {
        AuthApiService api = ApiClient.getRetrofit(context, true).create(AuthApiService.class);
        Call<BarberoResponse> call = api.listarBarberos();

        call.enqueue(new Callback<BarberoResponse>() {
            @Override
            public void onResponse(Call<BarberoResponse> call, Response<BarberoResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body().getData());
                } else {
                    callback.onError("Error al obtener barberos");
                }
            }

            @Override
            public void onFailure(Call<BarberoResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }

    public void crearBarbero(Context context, String nombre, BarberoOperacionCallback callback) {
        AuthApiService api = ApiClient.getRetrofit(context, true).create(AuthApiService.class);
        BarberoRequest nuevo = new BarberoRequest(nombre);

        api.crearBarbero(nuevo).enqueue(new Callback<BarberoResponse>() {
            @Override
            public void onResponse(Call<BarberoResponse> call, Response<BarberoResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body().getMessage());
                } else {
                    callback.onError("Error al crear barbero");
                }
            }

            @Override
            public void onFailure(Call<BarberoResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }

    public void actualizarBarbero(Context context, int id, String nuevoNombre, ActualizarCallback callback) {
        AuthApiService api = ApiClient.getRetrofit(context, true).create(AuthApiService.class);
        BarberoRequest request = new BarberoRequest(nuevoNombre);

        api.actualizarBarbero(id, request).enqueue(new Callback<BarberoSimpleResponse>() {
            @Override
            public void onResponse(Call<BarberoSimpleResponse> call, Response<BarberoSimpleResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body().getMessage());
                } else {
                    callback.onError("Error al actualizar barbero");
                }
            }

            @Override
            public void onFailure(Call<BarberoSimpleResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });

    }

    public void eliminarBarbero(Context context, int id, BarberoOperacionCallback callback) {
        AuthApiService api = ApiClient.getRetrofit(context, true).create(AuthApiService.class);

        api.eliminarBarbero(id).enqueue(new Callback<BarberoResponse>() {
            @Override
            public void onResponse(Call<BarberoResponse> call, Response<BarberoResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body().getMessage());
                } else {
                    callback.onError("Error al eliminar barbero");
                }
            }

            @Override
            public void onFailure(Call<BarberoResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/diamond/appcliente/viewmodel/GestionarBarberoViewModel.kt
package com.diamond.appcliente.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.diamond.barbershop.shared.dto.barbero.BarberoDto
import com.diamond.appcliente.repository.BarberoRepository
import com.shared.models.ui.state.UiState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class GestionarBarberoViewModel @Inject constructor(
    private val barberoRepository: BarberoRepository
) : ViewModel() {

    private val _barberos = MutableStateFlow<UiState<List<BarberoDto>>>(UiState.Idle)
    val barberos: StateFlow<UiState<List<BarberoDto>>> = _barberos.asStateFlow()

    private val _mensaje = MutableSharedFlow<String>()
    val mensaje: SharedFlow<String> = _mensaje.asSharedFlow()

    private val _crudError = MutableSharedFlow<String>()
    val crudError: SharedFlow<String> = _crudError.asSharedFlow()

    fun cargarBarberos() {
        viewModelScope.launch {
            _barberos.value = UiState.Loading
            try {
                _barberos.value = UiState.Success(barberoRepository.listarBarberos())
            } catch (e: Exception) {
                _barberos.value = UiState.Error(e.message ?: "Error desconocido")
            }
        }
    }

    fun crearBarbero(nombre: String) {
        viewModelScope.launch {
            try {
                _mensaje.emit(barberoRepository.crearBarbero(nombre))
                cargarBarberos()
            } catch (e: Exception) {
                _crudError.emit(e.message ?: "Error desconocido")
            }
        }
    }

    fun actualizarBarbero(id: Int, nuevoNombre: String) {
        viewModelScope.launch {
            try {
                _mensaje.emit(barberoRepository.actualizarBarbero(id, nuevoNombre))
                cargarBarberos()
            } catch (e: Exception) {
                _crudError.emit(e.message ?: "Error desconocido")
            }
        }
    }

    fun eliminarBarbero(id: Int) {
        viewModelScope.launch {
            try {
                _mensaje.emit(barberoRepository.eliminarBarbero(id))
                cargarBarberos()
            } catch (e: Exception) {
                _crudError.emit(e.message ?: "Error desconocido")
            }
        }
    }

    fun tipoHorarioToValor(tipoHorario: String?): Int = when (tipoHorario) {
        "MAÑANA" -> 1; "TARDE" -> 2; "NOCHE" -> 3; else -> -1
    }

    fun turnoReservaToValor(turno: String?): Int = when (turno) {
        "9 AM - 10 AM"  -> 1;  "10 AM - 11 AM" -> 2;  "11 AM - 12 PM" -> 3
        "12 PM - 1 PM"  -> 4;  "1 PM - 2 PM"   -> 5;  "2 PM - 3 PM"   -> 6
        "3 PM - 4 PM"   -> 7;  "4 PM - 5 PM"   -> 8;  "5 PM - 6 PM"   -> 9
        "6 PM - 7 PM"   -> 10; "7 PM - 8 PM"   -> 11; "8 PM - 9 PM"   -> 12
        "9 PM - 10 PM"  -> 13; else -> -1
    }
}
